### ISSUE TITLE
Type-narrowing Union with Literal using comparison to Literal variable

### ISSUE THREAD

**Issue by Garrett-R** [python/mypy#16465]
**Bug Report**

Strangely, type-narrowing for unions including a Literal works for `int` but not `datetime`

**To Reproduce**

[Mypy Playground](https://mypy-play.net/?mypy=1.7.0&python=3.11&gist=10e027bc9e23248edbf951aef68b1e17).  Copied here for convenience:

```python
from datetime import datetime
from typing import Final, Literal

Unspecified = Literal['unspecified']
UNSPECIFIED: Final[Unspecified] = 'unspecified'

def do_thing(x: datetime | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.utcnow())  # Unexpected error!

def do_thing_2(x: int | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.real)  # No prob
```

**Expected Behavior**

No errors.

**Actual Behavior**

Has errors:
```
main.py:9: error: Item "str" of "datetime | Literal['unspecified']" has no attribute "utcnow"  [union-attr]
```

**Your Environment**

- Mypy version used: 1.7.0
- Mypy command-line flags: <_none_>
- Mypy configuration options from `mypy.ini` (and other config files): <_none_>
- Python version used: 3.11

**Comment by vnmabus** [comment]
I would say that this is indeed a bug. It won't be a bug in the positive case (as a class that redefines `__eq__` MAY be equal to a string), but it should definitely narrow in the negative case. [Here](https://mypy-play.net/?mypy=1.7.0&python=3.11&gist=99a06d7b08a8d64b4dc08a5d11c8b465) you can see that the problem manifest itself depending on if you redefined `__eq__` or not:
```python
from datetime import datetime
from typing import Final, Literal

Unspecified = Literal['unspecified']
UNSPECIFIED: Final[Unspecified] = 'unspecified'

class KK():
    
    def kk(self):
        pass

class KK2():
    
    def kk(self):
        pass
    
    def __eq__(self, other):
        return False

def do_thing(x: datetime | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.utcnow())  # Unexpected error!

def do_thing_2(x: int | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.real)  # No prob
        
def do_thing_3(x: KK | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.kk)  # No prob
        
def do_thing_4(x: KK2 | Unspecified) -> None:
    if x != UNSPECIFIED:
        print(x.kk)  # Unexpected error!
```

**Comment by A5rocks** [comment]
Copying over my smaller reproducer from #18569:

```py
from typing import Literal

class A:
    def __eq__(self, other: object) -> bool:  # necessary
        return isinstance(other, A)

def f(v: A | Literal["text"]) -> A | None:
    if v == "text":
        reveal_type(v)  # N: Revealed type is "Union[__main__.A, Literal['text']]"
        return None
    else:
        reveal_type(v)  # N: Revealed type is "Union[__main__.A, Literal['text']]"
        return v
```
